File: scripts/lib/receita_wellness_filter.py
"""Receita CNAE wellness — segment match, dedup, tags (Python mirror of receitaWellnessFilter.ts)."""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]
DEFAULT_CONFIG = ROOT / "data" / "config" / "receita-cnae-segments.json"


def load_config(path: Path | None = None) -> dict[str, Any]:
    p = path or DEFAULT_CONFIG
    return json.loads(p.read_text(encoding="utf-8"))


def cnae_in_secondary(cnae: str, secundaria: str | None) -> bool:
    if not secundaria or not str(secundaria).strip():
        return False
    return cnae in [p.strip() for p in str(secundaria).split(",")]


def row_has_cnae(row: dict[str, Any], cnae: str) -> tuple[bool, str | None]:
    if str(row.get("cnae_fiscal_principal") or "") == cnae:
        return True, "principal"
    if cnae_in_secondary(cnae, row.get("cnae_fiscal_secundaria")):
        return True, "secundario"
    return False, None


def find_segment_hits(row: dict[str, Any], segments: list[dict[str, str]]) -> list[dict[str, Any]]:
    hits: list[dict[str, Any]] = []
    for i, seg in enumerate(segments):
        ok, match = row_has_cnae(row, seg["cnae"])
        if ok and match:
            hits.append({"order": i, "segment": seg, "match": match})
    return hits


def pick_winning_hit(hits: list[dict[str, Any]]) -> dict[str, Any] | None:
    if not hits:
        return None
    principals = [h for h in hits if h["match"] == "principal"]
    pool = principals if principals else hits
    return min(pool, key=lambda h: h["order"])


def apply_tag_rules(row: dict[str, Any], rules: list[dict[str, Any]]) -> list[str]:
    tags: set[str] = set()
    nome = str(row.get("nome_fantasia") or "")

    for rule in rules:
        if rule.get("match") == "nome_fantasia":
            pat = rule["pattern"].replace("(?i)", "", 1) if rule["pattern"].startswith("(?i)") else rule["pattern"]
            if re.search(pat, nome, re.IGNORECASE):
                tags.add(rule["id"])
        elif rule.get("match") == "cnae":
            ok, _ = row_has_cnae(row, rule["cnae"])
            if ok:
                tags.add(rule["id"])

    return sorted(tags)


def enrich_row(row: dict[str, Any], config: dict[str, Any]) -> dict[str, Any] | None:
    segments = config.get("segments") or []
    hits = find_segment_hits(row, segments)
    winner = pick_winning_hit(hits)
    if not winner:
        return None

    seg = winner["segment"]
    enriched = dict(row)
    enriched["cnae_match"] = winner["match"]
    enriched["cnae_segment"] = seg["id"]
    enriched["cnae_fiscal_matched"] = seg["cnae"]
    enriched["cnae_tags"] = apply_tag_rules(row, config.get("tags") or [])
    return enriched


def dedupe_rows(rows: list[dict[str, Any]], config: dict[str, Any]) -> list[dict[str, Any]]:
    by_cnpj: dict[str, dict[str, Any]] = {}
    segments = config.get("segments") or []

    for row in rows:
        cnpj = row.get("cnpj")
        if not cnpj:
            continue
        enriched = enrich_row(row, config)
        if not enriched:
            continue

        prev = by_cnpj.get(cnpj)
        if not prev:
            by_cnpj[cnpj] = enriched
            continue

        prev_w = pick_winning_hit(find_segment_hits(prev, segments))
        next_w = pick_winning_hit(find_segment_hits(row, segments))
        if not prev_w or not next_w:
            continue

        prev_score = 0 if prev_w["match"] == "principal" else 1
        next_score = 0 if next_w["match"] == "principal" else 1
        if next_score < prev_score or (
            next_score == prev_score and next_w["order"] < prev_w["order"]
        ):
            by_cnpj[cnpj] = enriched

    return list(by_cnpj.values())


def count_by_segment(rows: list[dict[str, Any]]) -> dict[str, int]:
    counts: dict[str, int] = {}
    for row in rows:
        seg = row.get("cnae_segment") or "?"
        counts[seg] = counts.get(seg, 0) + 1
    return counts

File: scripts/lib/receitaWellnessFilter.ts
/**
 * Receita CNAE wellness — segment match, dedup, tags.
 */
import fs from 'fs';
import path from 'path';

export type WellnessSegment = {
  id: string;
  cnae: string;
  label: string;
};

export type WellnessTagRule =
  | { id: string; match: 'nome_fantasia'; pattern: string }
  | { id: string; match: 'cnae'; cnae: string };

export type WellnessConfig = {
  version: number;
  segments: WellnessSegment[];
  segment_groups?: Record<string, string[]>;
  tags?: WellnessTagRule[];
};

export type WellnessEstabRow = {
  cnpj: string;
  nome_fantasia?: string | null;
  cnae_fiscal_principal?: string | null;
  cnae_fiscal_secundaria?: string | null;
  situacao_cadastral?: string | null;
  [key: string]: unknown;
};

export type WellnessEnrichedRow = WellnessEstabRow & {
  cnae_match: 'principal' | 'secundario';
  cnae_segment: string;
  cnae_fiscal_matched: string;
  cnae_tags: string[];
};

const ROOT = process.cwd();

export function loadWellnessConfig(configPath?: string): WellnessConfig {
  const p =
    configPath ?? path.join(ROOT, 'data/config/receita-cnae-segments.json');
  return JSON.parse(fs.readFileSync(p, 'utf8')) as WellnessConfig;
}

export function cnaeInSecondary(cnae: string, secundaria: string | null | undefined): boolean {
  if (!secundaria?.trim()) return false;
  const parts = secundaria.split(',').map((s) => s.trim());
  return parts.includes(cnae);
}

export function rowHasCnae(
  row: WellnessEstabRow,
  cnae: string,
): { hit: boolean; match: 'principal' | 'secundario' | null } {
  if (row.cnae_fiscal_principal === cnae) {
    return { hit: true, match: 'principal' };
  }
  if (cnaeInSecondary(cnae, row.cnae_fiscal_secundaria)) {
    return { hit: true, match: 'secundario' };
  }
  return { hit: false, match: null };
}

type SegmentHit = {
  order: number;
  segment: WellnessSegment;
  match: 'principal' | 'secundario';
};

export function findSegmentHits(row: WellnessEstabRow, segments: WellnessSegment[]): SegmentHit[] {
  const hits: SegmentHit[] = [];
  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i];
    const { hit, match } = rowHasCnae(row, seg.cnae);
    if (hit && match) hits.push({ order: i, segment: seg, match });
  }
  return hits;
}

export function pickWinningHit(hits: SegmentHit[]): SegmentHit | null {
  if (!hits.length) return null;
  const principals = hits.filter((h) => h.match === 'principal');
  const pool = principals.length ? principals : hits;
  return pool.reduce((best, cur) => (cur.order < best.order ? cur : best));
}

export function applyTagRules(row: WellnessEstabRow, rules: WellnessTagRule[] = []): string[] {
  const tags = new Set<string>();
  const nome = String(row.nome_fantasia ?? '');

  for (const rule of rules) {
    if (rule.match === 'nome_fantasia') {
      const pat = rule.pattern.replace(/^\(\?i\)/, '');
      if (new RegExp(pat, 'i').test(nome)) tags.add(rule.id);
      continue;
    }
    const { hit } = rowHasCnae(row, rule.cnae);
    if (hit) tags.add(rule.id);
  }

  return [...tags].sort();
}

export function enrichWellnessRow(
  row: WellnessEstabRow,
  config: WellnessConfig,
): WellnessEnrichedRow | null {
  const hits = findSegmentHits(row, config.segments);
  const winner = pickWinningHit(hits);
  if (!winner) return null;

  return {
    ...row,
    cnae_match: winner.match,
    cnae_segment: winner.segment.id,
    cnae_fiscal_matched: winner.segment.cnae,
    cnae_tags: applyTagRules(row, config.tags ?? []),
  };
}

/** Dedup by CNPJ — one row per estabelecimento. */
export function dedupeWellnessRows(
  rows: WellnessEstabRow[],
  config: WellnessConfig,
): WellnessEnrichedRow[] {
  const byCnpj = new Map<string, WellnessEnrichedRow>();

  for (const row of rows) {
    if (!row.cnpj) continue;
    const enriched = enrichWellnessRow(row, config);
    if (!enriched) continue;

    const prev = byCnpj.get(row.cnpj);
    if (!prev) {
      byCnpj.set(row.cnpj, enriched);
      continue;
    }

    const prevHits = findSegmentHits(prev, config.segments);
    const nextHits = findSegmentHits(row, config.segments);
    const prevWinner = pickWinningHit(prevHits);
    const nextWinner = pickWinningHit(nextHits);
    if (!prevWinner || !nextWinner) continue;

    const prevScore = prevWinner.match === 'principal' ? 0 : 1;
    const nextScore = nextWinner.match === 'principal' ? 0 : 1;
    if (
      nextScore < prevScore ||
      (nextScore === prevScore && nextWinner.order < prevWinner.order)
    ) {
      byCnpj.set(row.cnpj, enriched);
    }
  }

  return [...byCnpj.values()];
}

export function countBySegment(rows: WellnessEnrichedRow[]): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const row of rows) {
    counts[row.cnae_segment] = (counts[row.cnae_segment] ?? 0) + 1;
  }
  return counts;
}

export function segmentGroupFor(segmentId: string, config: WellnessConfig): string | null {
  for (const [group, ids] of Object.entries(config.segment_groups ?? {})) {
    if (ids.includes(segmentId)) return group;
  }
  return null;
}
